an-tao/drogon
# Drogon test framework

This PR introduces the new async Drogon Test (well, that's what I call it) testing framework. And fixes a few bugs discovered along the way.

* New async test framework
* Coroutine can return move-only objects
* AsyncTask cleanup
* Make `Set-Cookie`  comply with RFC
* Port unit tests
* `make test` runs unittest
* Move `webapp` and `webapp_test` into `integration_test_server` and `integration_test_client`
* Add web socket example
* Merge `lib/tests` and `unittests` into  a single unittest
* Test exception handling in coroutine based HTTP handler
* Remove GTest dependency
* ~~Make `BUILD_TESTING` an option as we doen't use CTest (Which is indirectly used by GTest) anymore.~~
   * Drogon Test now supports CTest

The new testing framework should reduce the complexity when writing new tests. As we no longer need to count how much tests are done. Now, the progress is tracked by reference counting through a testing context. Then automatically ends testing when all contexts are released.

Instead of writing
```c++
client->sendRequest(req,
    [req, isHttps](ReqResult result,
                   const HttpResponsePtr &resp) {
        if (result == ReqResult::Ok)
        {
            if (resp->getBody().length() == 4994)
            {
                outputGood(req, isHttps);
            }
            else
            {
                LOG_DEBUG << resp->getBody().length();
                LOG_ERROR << "Error!";
                exit(1);
            }
        }
        else
        {
            LOG_ERROR << "Error!";
            exit(1);
        }
    });
```

The same test becomes:

```c++
client->sendRequest(req,
    [req, TEST_CTX](ReqResult result,
                    const HttpResponsePtr &resp) {
        REQUIRE(result == ReqResult::Ok);
        CHECK(resp->getBody().length() == 4994);
    });
```

And generates informative messages when assertion failed.
```
❯ ./integration_test_client -r HttpTest
In test case HttpTest
↳ /home/marty/Documents/drogon/tests/integration_test/client/main.cpp:104  FAILED:
  CHECK(resp->getBody().length() == 4994)
With expansion
  0 == 4994


================================================================================
assertions: 127 | 126 passed | 1 failed
test cases:   1 |   0 passed | 1 failed
```

The syntax of Drogon Test follows [Catch2](https://github.com/catchorg/Catch2)'s, but with extension. `CHECK` simply checks. `(CO_)REQUIRE` checks and returns if the check fails. And `MANDATE` kills the entire process if failed. However, Drogon Test does not support fixtures due to the async nature.

All tests accepts a few parameters to assist in testing. `-r` forces to run a certain test (otherwise all tests run). And `-s` shows all successful assertions. A help message is available through `-h`.

Example: 

```
❯ ./lib/tests/unittest -h
A Drogon Test application:

Usage: ./lib/tests/unittest [options]
options:
    -r        Run a specific test
    -s        Print successful tests
    -l        List avaliable tests
    -h        Print this help message
❯ ./lib/tests/unittest -s -r HttpHeaderResponse
In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:22  PASSED:
  REQUIRE(resp != nullptr)

In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:24  PASSED:
  CHECK(resp->getHeader("Abc") == "abc")

In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:25  PASSED:
  CHECK(resp->getHeader("abc") == "abc")

In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:30  PASSED:
  CHECK(str.find("abc") != std::string::npos)

In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:35  PASSED:
  CHECK(str.find("abc") == std::string::npos)

In test case HttpHeaderResponse
↳ /home/marty/Documents/drogon/lib/tests/unittests/HttpHeaderTest.cc:36  PASSED:
  CHECK(resp->getHeader("abc") == "")


================================================================================
  All tests passed (6 assertions in 1 tests cases).

```

## Discovered (unresolved) bugs
* Memory leak when a huge amount or WS clients are spawned together
    * Try run `MultipleWsTest` with 1K clients with ASan on.
